protocol/testing/e2e/gov/perpetuals_test.go

```diff
@@ -0,0 +1,305 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	govtypesv1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
+	testapp "github.com/dydxprotocol/v4-chain/protocol/testutil/app"
+	"github.com/dydxprotocol/v4-chain/protocol/testutil/constants"
+	perptypes "github.com/dydxprotocol/v4-chain/protocol/x/perpetuals/types"
+	"github.com/stretchr/testify/require"
+)
+
+// This tests `MsgUpdateParams` in `x/perpetuals`.
+func TestUpdatePerpetualsModuleParams(t *testing.T) {
+	tests := map[string]struct {
+		msg                      *perptypes.MsgUpdateParams
+		expectedProposalStatus   govtypesv1.ProposalStatus
+		expectSubmitProposalFail bool
+	}{
+		"Success": {
+			msg: &perptypes.MsgUpdateParams{
+				Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+				Params: perptypes.Params{
+					FundingRateClampFactorPpm: 123_456,
+					PremiumVoteClampFactorPpm: 123_456_789,
+					MinNumVotesPerSample:      15,
+				},
+			},
+			expectedProposalStatus: govtypesv1.ProposalStatus_PROPOSAL_STATUS_PASSED,
+		},
+		"Fail: invalid authority": {
+			msg: &perptypes.MsgUpdateParams{
+				Authority: authtypes.NewModuleAddress(perptypes.ModuleName).String(),
+				Params: perptypes.Params{
+					FundingRateClampFactorPpm: 100,
+					PremiumVoteClampFactorPpm: 100,
+					MinNumVotesPerSample:      15,
+				},
+			},
+			expectSubmitProposalFail: true,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			tApp := testapp.NewTestAppBuilder().WithGenesisDocFn(func() (genesis types.GenesisDoc) {
+				genesis = testapp.DefaultGenesis()
+				testapp.UpdateGenesisDocWithAppStateForModule(
+					&genesis,
+					func(genesisState *govtypesv1.GenesisState) {
+						genesisState.Params.VotingPeriod = &testapp.TestVotingPeriod
+					},
+				)
+				// Initialize perpetuals module with params that are different from the proposal.
+				testapp.UpdateGenesisDocWithAppStateForModule(
+					&genesis,
+					func(genesisState *perptypes.GenesisState) {
+						genesisState.Params = perptypes.Params{
+							FundingRateClampFactorPpm: tc.msg.Params.FundingRateClampFactorPpm + 1,
+							PremiumVoteClampFactorPpm: tc.msg.Params.PremiumVoteClampFactorPpm + 2,
+							MinNumVotesPerSample:      tc.msg.Params.MinNumVotesPerSample + 3,
+						}
+					},
+				)
+				return genesis
+			}).WithTesting(t).Build()
+			ctx := tApp.InitChain()
+			initialParams := tApp.App.PerpetualsKeeper.GetParams(ctx)
+
+			// Submit and tally governance proposal that includes `MsgUpdateParams`.
+			ctx = testapp.SubmitAndTallyProposal(
+				t,
+				ctx,
+				&tApp,
+				[]sdk.Msg{tc.msg},
+				tc.expectSubmitProposalFail,
+				tc.expectedProposalStatus,
+			)
+
+			// If governance proposal is supposed to fail submission, verify that perpetuals
+			// params match the ones before proposal submission.
+			if tc.expectSubmitProposalFail {
```

Comment:
Sorry to revisit this after you merged, but... can you add this check-original-state clause to the bridge it tests? https://github.com/dydxprotocol/v4-chain/pull/425/files

Reply:
will add it in upcoming PR that adds more bridge test cases!